### PR TITLE
serializeIdsInAttributes

### PR DESCRIPTION
**Pull request by jnaglick**
This adds a metadata option to hasMany relationships called `serializeIdsInAttributes`

When this option is specified as `true`, a property with key `whatever_ids` having a value equal to the mapped `id` properties of the current `Whatevers` will be added to the attributes property of the top-level entity being serialized.

The purpose of this is to do with an option what the intention of https://github.com/jsonapi-suite/ember-data-extensions/pull/21 is to be used for (IE, `child_ids` will be added to the parent's attributes, allowing the user to simply manipulate the `children` collection as they'd expect, rather than having to mess with `markForDeletion` and etc)

Some issues:

(1) This option should probably also disclude `Whatevers` from `relationships`

(2) This currently *only* works for relationships of the top-level entity, not "sub" relationships of its children, because I'm only passing the relationships to `attributesFor` within the primary `serialize` method and not also the `jsonapiPayload` method

I could probably address these issues but the build/test server has 'randomly' broken on me about 6 times today during development and I'm exhausted with re-cloning this repo 😊

### CODE REVIEW

File: addon/mixins/nested-relations.js [hunk 1]
```diff
@@ -203,7 +214,9 @@ export default Mixin.create({
 
       let adapterOptions = snapshot.adapterOptions || {};
 
-      let attributes = attributesFor(snapshot.record);
+      let relationships = relationshipsDirective(adapterOptions.relationships);
+
+      let attributes = attributesFor(snapshot.record, relationships);
```
Comment (onumossn) [inline addon/mixins/nested-relations.js]: `attributesFor` is getting only the subsets of relationships that are passed in for serialization. This means the `attributes` can only contain relationship ids of things that are normally serialized, so `(1)` from https://github.com/jsonapi-suite/ember-data-extensions/pull/23#issue-331124621 cannot be done.

**Review by onumossn — COMMENTED** [review]
1. This isn't very useful for us without `(2)` from https://github.com/jsonapi-suite/ember-data-extensions/pull/23#issue-331124621 addressed.
2. I really disagree with this method. I had done something like this in an earlier PR, but decided that this library shouldn't contain this special code for our personal use cases. The library should stay generic and meaningful to everyone and not just us and our applications. If everyone else is ok with this method, I am fine with it I guess.

File: addon/mixins/nested-relations.js [hunk 2]
```diff
@@ -51,6 +51,17 @@ const attributesFor = function(record) {
     }
   });
 
+  if (relationshipsHash !== undefined) {
+    iterateRelations(record, relationshipsHash, (name, kind, related, subRelations, manyToManyDeleted) => {
+      let metadata = record.relationshipFor(name);
+
+      if (metadata.options && metadata.options.serializeIdsInAttributes) {
+        let keyName = `${serializer.keyForAttribute(name)}_ids`;
```
Comment (onumossn) [inline addon/mixins/nested-relations.js]: This only handles `hasMany` relationships correctly. It doesn't work with `belongsTo`.

File: addon/mixins/nested-relations.js (hunk 1 above)
Comment (jnaglick) [reply addon/mixins/nested-relations.js]: We should not be discluding from relationships within this method. It should be done within the method that serializes relationships

File: addon/mixins/nested-relations.js (hunk 2 above)
Comment (jnaglick) [reply addon/mixins/nested-relations.js]: Yeah it is [currently] intentionally specific to hasMany relationships

**Comment by jnaglick** [comment]
> I really disagree with this method. I had done something like this in an earlier PR, but decided that this library shouldn't contain this special code for our personal use cases. The library should stay generic and meaningful to everyone and not just us and our applications.

I see where you're coming from however since this is essentially to workaround problems in a ruby gem that is open to the public I'd say it's not "special code for our personal use cases."

However, an alternative approach would be to take what you did (adding a 'hook' within `attributesFor`) and implement what I have in a Mixin in our codebase (because it's in literally the same place as your hook.)

However we'd also need more hooks to address the issues I mentioned in my top-level comment

**Comment by onumossn** [comment]
I agree with the alternate approach. I don't think `(1)` requires additional hooks, `(2)` doesn't necessarily need it either but would be convenient to have.

I still think insuring disclusion can be handled by the interface the server provides. Its like saying what happens if you mark something to be destroyed and then mark it to be disassociated or some other set of methods which the library didn't seem to expect would be called before serialization. I think the ensured disclusion is a nice to have, but not necessary.

**Pull request closed without merging**